File: Napakalaki RUBY/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"
require_relative "monster.rb"
require_relative "card_dealer.rb"
require_relative "player.rb"
require_relative "bad_consequence.rb"
require_relative "combat_result.rb"
require_relative "dice.rb"
require_relative "treasure.rb"
require_relative "treasure_kind.rb"
require_relative "prize.rb"
require_relative "cultist.rb"
require_relative "cultist_player.rb"
require_relative "specific_bad_consequence.rb"
require_relative "numeric_bad_consequence.rb"
require_relative "death_bad_consequence.rb"

module NapakalakiGame
  
  class Napakalaki
  
  include Singleton
  
  private
    
    def initPlayers(names)
      
      @players = Array.new
      
      names.each do |nombre|
        @players << Player.newPlayer(nombre)
      end
    end
    
    def nextPlayer()
      
      if(@currentPlayer == nil)
        numeroAleatorio = rand(0..(@players.size-1))
        indice = numeroAleatorio
        
      elsif (@currentPlayer == @players[@players.size-1])
        indice = 0
      
      else
        indice = @players.index(@currentPlayer) + 1;
      end
    
      nextPlayer = @players[indice]
      @currentPlayer = nextPlayer
      
      return nextPlayer
    end
    
    def nextTurnAllowed()
      allow = false
      
      if(@currentPlayer == nil)
        allow = true
      elsif (@currentPlayer.validState())
        allow = true
      end
      
      return allow
    end
    
    def setEnemies()
      
      @players.each do |p|
        numeroAleatorio = rand(0..(@players.size - 1))
        enemy = @players[numeroAleatorio]
        while(enemy == p)
          numeroAleatorio = rand(0..(@players.size - 1))
          enemy = @players[numeroAleatorio]
        end
        
        p.setEnemy(enemy)
      end
      
    end
    
  public
    
    def getCurrentPlayer()
      @currentPlayer
    end
    
    def getCurrentMonster()
      @currentMonster
    end
    
    def developCombat()
      combatResult = @currentPlayer.combat(@currentMonster)
      @dealer.giveMonsterBack(@currentMonster)
      
      if combatResult == CombatResult::LOSEANDCONVERT
        cultist = @dealer.nextCultist()
        
        cultistPlayer = CultistPlayer.new(@currentPlayer, cultist)
        
        @players.each do |p|
          if(p.getEnemy() == @currentPlayer)
            p.setEnemy(cultistPlayer)
          end
        end
        
        indice = @players.index(@currentPlayer);
        @players.delete_at(indice)
        @players.insert(indice, cultistPlayer)
        
        @currentPlayer = cultistPlayer
      end
      
      return combatResult
    end
    
    def discardVisibleTreasures(treasures)
      treasures.each do |treasure|
        @currentPlayer.discardVisibleTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end
    
    def discardHiddenTreasures(treasures)
      treasures.each do |treasure|
        @currentPlayer.discardHiddenTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end
    
    def makeTreasuresVisible(treasures)
      treasures.each do |t|
        @currentPlayer.makeTreasureVisible(t)
      end
    end
    
    def initGame(players)
      @dealer = CardDealer.instance
      initPlayers(players)
      setEnemies()
      @dealer.initCards()
      nextTurn()
    end
    
    def nextTurn()
      stateOK = nextTurnAllowed()
      
      if(stateOK)
        @currentMonster = @dealer.nextMonster()
        @currentPlayer = nextPlayer()
        
        dead = @currentPlayer.isDead()
        
        if(dead)
          @currentPlayer.initTreasures()
        end
      end
      
      return stateOK
    end
    
    def endOfGame(result)
      terminado = false
      
      if (result == CombatResult::WINGAME)
        terminado = true
      end
      
      return terminado
    end

end
end

File: Napakalaki JAVA : GUI/src/napakalaki_java/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki_java;
import java.util.*;

/**
 *
 * @author jumacasni
 */
public class Napakalaki {
    
    private static Napakalaki instance = null;
    private Monster currentMonster;
    private CardDealer dealer;
    private ArrayList<Player> players;
    private Player currentPlayer;
    private Napakalaki() {
    }
    
    private void initPlayers(String[] names){
        players = new ArrayList();
        
        for(String nombre : names)
            players.add(new Player(nombre));
    }
    
    private Player nextPlayer(){
        
        Player nextPlayer;
        int numeroAleatorio;
        int indice;
        
        if(currentPlayer == null){
            numeroAleatorio = (int) (Math.random()*(players.size()));
            indice = numeroAleatorio;
        }
        
        else{
            if(currentPlayer == players.get(players.size()-1))
                indice = 0;
            
            else
                indice = players.indexOf(currentPlayer) + 1;
            
        }
        
        nextPlayer = players.get(indice);
        currentPlayer = nextPlayer;
        
        return nextPlayer;
    }
    
    private boolean nextTurnAllowed(){
        boolean allow = false;
        
        if(currentPlayer == null)
            allow = true;
        else if(currentPlayer.validState())
            allow = true;
        
        return allow;    
    }
    
    private void setEnemies(){
        int numeroAleatorio;
        Player enemy;
        
        for(Player p : players){
            numeroAleatorio = (int) (Math.random()*(players.size()));
            enemy = players.get(numeroAleatorio);
                while(enemy == p){
                numeroAleatorio = (int) (Math.random()*(players.size()));
                enemy = players.get(numeroAleatorio);
            }
            
            p.SetEnemy(enemy);
        }
        
    }
    
    public static Napakalaki getInstance() {
        if (instance == null)
                instance = new Napakalaki();
        
        return instance;
    }
    
    public CombatResult developCombat(){
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        
        if(combatResult == CombatResult.LOSEANDCONVERT){
            Cultist card = dealer.nextCultist();
            CultistPlayer sectario = new CultistPlayer(currentPlayer, card);
            
            // Reemplazar atributo enemy:
            for(Player p : players)
                if(p.getEnemy() == currentPlayer)
                    p.SetEnemy(sectario);
            
            // Reemplazar de la lista de Players:
            int pos = players.indexOf(currentPlayer);
            players.set(pos,sectario);
            
            // Reemplazar CurrentPlayer::
            currentPlayer = sectario;
        }
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for(Treasure treasure : treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);        
        }
        
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        for(Treasure treasure : treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);        
        }
        
    }
    
    public void makeTreasuresVisible (ArrayList<Treasure> treasures){
        for(Treasure t : treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame (ArrayList<String> players){
        initPlayers(players.toArray(new String[players.size()]));
        setEnemies();
        dealer = CardDealer.getInstance();
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK = nextTurnAllowed();
        
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            boolean dead = currentPlayer.isDead();
            
            if(dead){
                currentPlayer.initTreasures();
            }
        }
        
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        boolean end = false;
        
        if(result == CombatResult.WINGAME)
            end = true;
        
        return end;
    }
}
